sections/2-language-rules.md

```diff
@@ -645,3 +645,54 @@ package object dsl {
 }
 
 ```
+
+
+### 2.19 MUST NOT use `Seq.head`
+
+You might be tempted to this:
+
+```scala
+val userList: List[User] = ???
+
+// ....
+val firstName = userList.head.firstName
+```
+
+Don't ever do this, as this will throw `NoSuchElementException` if the sequence is empty.
+
+Alternatives:
+
+1. using `Seq.headOption` possibly combined with `getOrElse` or pattern matching
+
+    Example:
+    
+    ```scala
+    val firstName = userList.headOption match {
+        case Some(user) => user.firstName
+        case _ => "Unknown"
+      }
+    ```
+
+2. using pattern matching with the extractor `+:` 
+
+    Example:
+    
+    ```scala
+    val firstName = userList match {
+     case head +: _ => head.firstName
```

Comment:
TBH I don't like this sample. It's one thing to get the `head`, however this extractor gets the `tail` as well, but then discards it.

To understand why that is a problem, consider this sample:

```scala
val bigArray = (0 until 10000000).toArray

bigArray match {
  case head +: _ => ???
}
```

What this will do is to create another array of 9999999 integers.

For most collections `head` is an `O(1)` operation, given that you only need to keep an internal pointer for it, however there are really few collections that have a `tail` that is `O(1)` or at least an `O(log2 n)`.

In fact I can only think of only one persistent data-structure that has an `O(1)` `tail` and that is `List`. And then there's `Stream`, which is like a broken `List` that should never be used, and then you have the "amortized constant" one in `Queue`, but that's because it uses `List` internally. And `Vector` which claims to be "effectively constant", aka "pretty bad if you actually care".

So not only would I delete this `+:` extractor, but I'd create a best practice rule out of it.

Reply:
Agreed: head/tail decomposition should only be used, when both `head` and `tail` are `O(1)`. Same goes for init/last decomposition. There SHOULD be a rule for that ;)

Reply:
BTW: `+:` is also head/tail decomposition and `:+` is also init/last decomposition.